### PR TITLE
Checking error when calling output

### PR DESCRIPTION
**Pull request by alecha**
We've noticed that the error returned by the `Write` function is not used in any way, and this was causing some issues when the connection wasn't alive.

This fix will try to connect again when an error is arised (the fmt logs are just for debug, they are of course removable)

### CODE REVIEW

**Comment by bsphere** [comment]
Can you please make sure that tests are passing?

File: le.go [hunk 1]
```diff
@@ -138,9 +138,21 @@ func (logger *Logger) Flags() int {
...
 // Output does the actual writing to the TCP connection
 func (logger *Logger) Output(calldepth int, s string) error {
-	_, err := logger.Write([]byte(s))
-
-	return err
+	var err error
+	for {
+		_, err = logger.Write([]byte(s))
+		if err != nil {
+			fmt.Printf("go_le error: %v\n\tTrying to recover...", err)
```
Comment (MahlerFive) [inline le.go]: You shouldn't really be printing inside a library call.. same goes for the below calls to fmt.Printf()

File: le.go [hunk 2]
```diff
@@ -138,9 +138,21 @@ func (logger *Logger) Flags() int {
 
 // Output does the actual writing to the TCP connection
 func (logger *Logger) Output(calldepth int, s string) error {
-	_, err := logger.Write([]byte(s))
-
-	return err
+	var err error
+	for {
```
Comment (MahlerFive) [inline le.go]: This has the potential to be an infinite loop if the connection is dead and unable to recover (eg. server is down).
Retry once or with an exponential backoff maybe?

File: le.go [hunk 3]
```diff
@@ -138,9 +138,21 @@ func (logger *Logger) Flags() int {
 
 // Output does the actual writing to the TCP connection
 func (logger *Logger) Output(calldepth int, s string) error {
```
Comment (MahlerFive) [inline le.go]: I'm thinking the reconnect logic would make more sense in Write() where it already does a connection test.

**Comment by bsphere** [comment]
thanks for the fixes and improvements.. can you please make the tests pass so I can merge this PR ?

File: le.go (hunk 2 above)
Comment (alecha) [inline le.go]: you are right, I will update my PR

**Comment by alecha** [comment]
I've removed fmt statements but a test fails because of the the tcp connection timeout

**Comment by MahlerFive** [comment]
Thinking about it more - wouldn't it make more sense that the call ends up returning an error to the caller if it fails to reconnect + write, and the caller can decide if and how often they want to retry?

**Comment by alecha** [comment]
yes it's an option, otherwise we can let the user change these values (in a sort of settings object), and we can even use this package for retrying (https://github.com/matryer/try)

**Comment by bsphere** [comment]
@alecha what's the status of this?

**Comment by alecha** [comment]
Sorry I didn't push.  I can push the change so the error is returned to the caller so he can implement the reconnection login on his own, but my idea was to give a out-of-the-box mechanism (maybe configurable with a `MaxConnectionRetries` parameter)

File: le.go [hunk 4]
```diff
@@ -138,9 +138,22 @@ func (logger *Logger) Flags() int {
...
-	_, err := logger.Write([]byte(s))
-
-	return err
+	var (
+		err        error
+		waitPeriod = time.Millisecond
+	)
+	for {
+		_, err = logger.Write([]byte(s))
+		if err != nil {
```
Comment (azumads) [inline le.go]: even the tcp is broken, err still can be nil, so it's not work.

**Comment by sascha-andres** [comment]
We have similar errors when using docker with the logentries logging driver (which uses this library). Any idea when this PR is complete?

**Comment by marcobrinkmann** [comment]
Any news on this?

File: le.go (hunk 4 above)
Comment (migueleliasweb) [reply le.go]: If net.SetWriteDeadline is set err would not be nil as the request would timeout. Right ?

**Comment by rfink** [comment]
@migueleliasweb I know it's been ages on this - but do you remember if you also sometimes experienced the "connect" hanging indefinitely?

**Pull request merged**